packages/atauth/src/run_onboard_command.c

```diff
@@ -0,0 +1,187 @@
+  }
+
+  // connect to atServer via cram auth
+  atclient atclient;
+  atclient_init(&atclient); // free me
+
+  atclient_authenticate_options auth_opts;
+  atclient_authenticate_options_init(&auth_opts); // free me
+
+  ret =
+      atclient_utils_find_atserver_address(root_domain, 64, atsign, &auth_opts.atserver_host, &auth_opts.atserver_port);
+  if (ret != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to lookup the atserver host & port\n");
+    goto free_auth_options;
+  }
+
+  ret = atclient_cram_authenticate(&atclient, atsign, cram_key, &auth_opts);
+  if (ret != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to cram authenticate to the atserver\n");
+    goto free_auth_options;
+  }
+
+  // generate new apkam keys
+  struct atauth_apkam_keys apkam_keys;
+  atauth_apkam_keys_init(&apkam_keys); // free me
+
+  ret = atauth_apkam_keys_generate_all(&apkam_keys);
+  if (ret != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to generate apkam keys\n");
+    goto free_apkam_keys;
+  }
+
+  // send enroll request
+  atauth_enroll_params_t ep = {
+      .app_name = ATAUTH_DEFAULT_FIRST_APP_NAME,
+      .device_name = ATAUTH_DEFAULT_FIRST_DEVICE_NAME,
+      .apkam_public_key = apkam_keys.pkam_public_key_base64,
+      .encrypted_default_encryption_private_key = apkam_keys.encrypted_encrypt_private_key_base64,
+      .encrypted_default_encryption_private_key_iv = apkam_keys.encrypted_encrypt_private_iv_base64,
+      .encrypted_self_encryption_key = apkam_keys.encrypted_self_encryption_key_base64,
+      .encrypted_self_encryption_key_iv = apkam_keys.encrypted_self_encryption_key_iv_base64,
+  };
+
+  struct enroll_response response;
+  ret = atauth_send_enroll_request(&atclient, &ep, atauth_apkam_request, &response);
+  if (ret != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atauth_send_enroll_request: %d\n", ret);
+    goto free_apkam_keys;
+  }
+
+  // convert apkam_keys into an atkeys
+  atclient_atkeys *atkeys;
+  ret = atauth_apkam_keys_create_atkeys(&apkam_keys, &atkeys);
+  if (ret != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to create the atkeys from the generated apkam keys\n");
+    goto free_enroll_response;
+  }
+
+  ret = atclient_atkeys_set_enrollment_id(atkeys, response.enrollment_id, strlen(response.enrollment_id));
+  if (ret != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to set enrollment_id in atkeys\n");
+    goto free_atkeys;
+  }
+
+  // disconnect
+  atclient_connection_disconnect(&atclient.atserver_connection);
+  atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_DEBUG, "Terminated cram instance of atserver connection\n");
+
+  // pkam auth
+  ret = atclient_pkam_authenticate(&atclient, atsign, atkeys, &auth_opts, NULL);
+  if (ret != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to pkam auth to the atServer\n");
+    goto free_atkeys;
+  }
+
+  // write atKeys to atKeys file
```

Comment:
Retries have not been implemented yet. I want to ensure this builds to our desired targets and finish enroll first